### ISSUE TITLE
Using with an iPhone

### ISSUE THREAD

**Issue by victorarthur** [KLUSEK/BMW-RPi-iBUS#4]
I could make it discoverable on my Mac but neither on iPhone 6 & 7 "sees" Raspberry Pi (using the built-in Zero W BT module) btw I'm building this for an X3 (e83) 2004-2011 once I'm done I will record a video in English for the project.

Do you have an idea what may cause this issue?

![screenshot 2018-03-27 21 03 22](https://user-images.githubusercontent.com/3355074/37988915-6acce1dc-3202-11e8-9886-cdb07f99a2c3.jpg)

**Comment by KLUSEK** [comment]
How about any phone?
I'm asking because I had a problem to connect from my PC Laptop, but no problem with androids or iphones

**Comment by victorarthur** [comment]
I tried with:

iMac (OK)
MacBook Air (OK)
MacbookPro (OK)
iPhone 5 (Failed)
iPhone 6 (Failed)
iPhone 7 (Failed)

However, once the device connected even on desktop after 10-15 secs it drops the connection, probably because I haven't connected the iBUS to the car yet.

![img_2056 heic](https://user-images.githubusercontent.com/3355074/37989706-85676146-3204-11e8-9ff0-d14ab02a310b.jpg)


I have no access to Android devices right now to test,

**Comment by KLUSEK** [comment]
Really hard to say.
Please try to start the script not as a demon to see output. If there is any exception you should see what is it.

As you probably know at this point I was doing nothing with bluetooth stack but "control" it via dbus.
So if this is not working with iOS next step should be googling for "Bluez Audio iOS problem" or similar ;)

Good luck

**Comment by KLUSEK** [comment]
PS.

**Comment by victorarthur** [comment]
Thanks! It seems iOS doesn't support not certified BT stacks natively. It works with various apps but not with the native BT scanner. So technically this project cannot be implemented for iOS. :(

**Comment by KLUSEK** [comment]
Weird ...

Some people have problem with iOS but some don't (https://gist.github.com/oleq/24e09112b07464acbda1)

Have you tried put my scripts aside and connect/pair/create audio sink manualy from console and check if iPhone works then?

PS. In the weekend I will try to check it with my girlfriend's iOS11

**Comment by victorarthur** [comment]
Thanks for this article getting there 👍 

The script needs to be modified for iOS, so it auto-connects. The point is you need to start a paring from rPi to iOS and create a new file within:

```
pi@raspberrypi:~ $ cat /etc/bluetooth/audio.conf
[General]
Class = 0x20041C
Enable = Source,Sink,Media,Socket
```

Now I have to figure out how to do this automatically on startup with any devices. Once it's done I will send a PR so others can use with iOS this great project as well ;) 

![img_2058](https://user-images.githubusercontent.com/3355074/38034177-fd940cbc-32a1-11e8-9206-f2394f4978cd.PNG)


![img_2057](https://user-images.githubusercontent.com/3355074/38034187-02cfb03c-32a2-11e8-9762-c14f97631fab.PNG)

**Comment by KLUSEK** [comment]
Great !

But I thought that /etc/bluetooth/audio.conf is obsolete and those settings are overided by Bluez.
Are you sure you need them or you need just initialize connection from RPi to iOS?

Actually in my code ... once you connect to device its mac address is being saved in file ".device" [defined here](https://github.com/KLUSEK/BMW-RPi-iBUS/blob/89a3ef971fa85a11213e760182edb643e612d8e0/bluetooth.py#L15) and then you can press "Voice button" on stearing wheel then RPi will try to connect to device with this mac address ([reconnect function](https://github.com/KLUSEK/BMW-RPi-iBUS/blob/89a3ef971fa85a11213e760182edb643e612d8e0/bluetooth.py#L167))

**Comment by KLUSEK** [comment]
Additionaly my scripts tries to set agent to NoInputNoOutput value and prompt for PIN should not be displayed, but I dont know how this affects to iOS

https://github.com/KLUSEK/BMW-RPi-iBUS/blob/89a3ef971fa85a11213e760182edb643e612d8e0/bluetooth.py#L63

**Comment by victorarthur** [comment]
I noticed a weird thing with the PulseAudio service as well.

```
[Unit]
Description=Pulse Audio

[Service]
Type=simple
ExecStart=/usr/bin/pulseaudio --system --disallow-exit --disable-shm --exit-idle-time=-1

[Install]
WantedBy=multi-user.target
```

If it runs in System mode it doesn't make it discoverable on iOS but on Desktop it discoverable. if I simply start it manually like `pulseaudio --start` it works perfectly on iOS as well. The issue here this way I cannot make it auto-starts.

I guess it may be because some device like the USB audio sound card or the BT adapter is not reachable at boot. But i don't know (yet) how to overcome this issue.

**Comment by KLUSEK** [comment]
Try to edit https://github.com/KLUSEK/BMW-RPi-iBUS/blob/89a3ef971fa85a11213e760182edb643e612d8e0/etc/systemd/system/bmw-agent.service#L10

and replace WantedBy to After

**Comment by pascal20997** [comment]
I´m facing the same issue using an iPhone Xs and Raspberry Pi 3+ with Raspbian Stretch. Manually starting the service works fine as described by @victorarthur. I experimented a little bit and got it working running the service as user pi and without --system flag.
EDIT: Added requirements to prevent exceptions because services maybe starting too late.

/etc/systemd/system/pulseaudio.service:
```
[Unit]
Description=Pulse Audio
Requires=pulseaudio.service dbus.service

[Service]
User=pi
Type=simple
ExecStart=/usr/bin/pulseaudio --disallow-exit --disable-shm --exit-idle-time=-1

[Install]
WantedBy=multi-user.target
```
